app/uk/gov/hmrc/tai/model/domain/NpsIabdSummaryHipReads.scala

```diff
@@ -30,27 +30,48 @@ object NpsIabdSummaryHipReads {
   }
 
   def totalLiabilityIabds(json: JsValue, subPath: String, categories: Seq[String]): Seq[NpsIabdSummary] = {
-    val pf: PartialFunction[JsValue, NpsIabdSummary] = {
+
+    val parseSummary: PartialFunction[JsValue, NpsIabdSummary] = {
       case json if (json \ "type").asOpt[(String, Int)](readsTypeTuple).isDefined =>
         val fullType = (json \ "type").as[String]
         val (description, componentType) = parseTypeOrException(fullType)
         val employmentId = (json \ "employmentSequenceNumber").asOpt[Int]
         val amount = (json \ "amount").asOpt[BigDecimal].getOrElse(BigDecimal(0))
         NpsIabdSummary(componentType, employmentId, amount, description)
     }
-    val list1 = categories
-      .flatMap(category =>
-        (json \ "totalLiabilityDetails" \ category \ subPath \ "summaryIABDDetailsList").asOpt[JsArray]
-      )
-      .flatMap(_.value) collect pf
 
-    val list2 = categories
-      .flatMap(category =>
-        (json \ "totalLiabilityDetails" \ category \ subPath \ "summaryIABDEstimatedPayDetailsList").asOpt[JsArray]
-      )
-      .flatMap(_.value) collect pf
+    def extractItems(listName: String): Seq[NpsIabdSummary] =
+      categories
+        .flatMap(category => (json \ "totalLiabilityDetails" \ category \ subPath \ listName).asOpt[JsArray])
+        .flatMap(_.value)
+        .collect(parseSummary)
+
+    val allItems = extractItems("summaryIABDDetailsList") ++ extractItems("summaryIABDEstimatedPayDetailsList")
+
+    checkForDuplicates[NpsIabdSummary](allItems, (item: NpsIabdSummary) => (item.employmentId, item.componentType))
 
-    list1 ++ list2
+    allItems
+  }
+
+  def checkForDuplicates[T](items: Seq[T], uniqueKey: T => (Option[Int], Int)): Unit = {
+    val duplicates = items
+      .groupBy(uniqueKey)
+      .collect { case (key, items) if items.size > 1 => key }
+      .toSeq
+      .sorted
+
+    if (duplicates.nonEmpty) {
+      val duplicateMessages = duplicates.distinct
+        .map { case (employmentId, componentType) =>
+          s"employmentSequenceNumber: $employmentId and componentType: $componentType"
```

Comment:
This isn't generic and reusable if we refer to specific fields here. Either make it non-generic in which case you don't need to bother passing in the function or make it generic in which case I'd put it into a different file in utils package